Sources/PubNub/EventEngine/Subscribe/SubscribeTransition.swift

```diff
@@ -0,0 +1,366 @@
+//
+//  PubNub Real-time Cloud-Hosted Push API and Push Notification Client Frameworks
+//  Copyright © 2023 PubNub Inc.
+//  https://www.pubnub.com/
+//  https://www.pubnub.com/terms
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+class SubscribeTransition: TransitionProtocol {
+  typealias State = (any SubscribeState)
+  typealias Event = Subscribe.Event
+  typealias Invocation = Subscribe.Invocation
+  
+  private func willExit(from state: State) -> [EffectInvocation<Invocation>] {
+    switch state {
+    case is Subscribe.HandshakingState:
+      return [.cancel(.handshakeRequest)]
+    case is Subscribe.HandshakeReconnectingState:
+      return [.cancel(.handshakeReconnect)]
+    case is Subscribe.ReceivingState:
+      return [.cancel(.receiveMessages)]
+    case is Subscribe.ReceiveReconnectingState:
+      return [.cancel(.receiveReconnect)]
+    default:
+      return []
+    }
+  }
+  
+  private func canTransition(from state: State, dueTo event: Event) -> Bool {
+    switch event {
+    case .handshakeSucceess(_):
+      return state is Subscribe.HandshakingState
+    case .handshakeFailure(_):
+      return state is Subscribe.HandshakingState
+    case .handshakeReconnectSuccess(_):
+      return state is Subscribe.HandshakeReconnectingState
+    case .handshakeReconnectFailure(_):
+      return state is Subscribe.HandshakingState || state is Subscribe.HandshakeReconnectingState
+    case .handshakeReconnectGiveUp(_):
+      return state is Subscribe.HandshakeReconnectingState
+    case .receiveSuccess(_,_):
+      return state is Subscribe.ReceivingState
+    case .receiveFailure(_):
+      return state is Subscribe.ReceivingState
+    case .receiveReconnectSuccess(_,_):
+      return state is Subscribe.ReceiveReconnectingState
+    case .receiveReconnectFailure(_):
+      return state is Subscribe.ReceivingState || state is Subscribe.ReceiveReconnectingState
```

Comment:
Should `ReceivingState` be in this check? `HANDSHAKE_RECONNECT_FAILURE` triggered only when we are in reconnecting state.